src/game/client/shadow/w3dprojectedshadow.cpp

```diff
@@ -0,0 +1,1581 @@
+        }
+    }
+}
+#endif
+
+W3DProjectedShadow::~W3DProjectedShadow()
+{
+    Ref_Ptr_Release(m_shadowProjector);
+
+    for (int i = 0; i < 1; i++) {
+        Ref_Ptr_Release(m_shadowTexture[i]);
+    }
+}
+
+W3DProjectedShadowManager::W3DProjectedShadowManager() :
+    m_shadowList(nullptr),
+    m_decalList(nullptr),
+    m_numDecalShadows(0),
+    m_numProjectionShadows(0),
+    m_W3DShadowTextureManager(nullptr),
+    m_shadowCamera(nullptr),
+    m_shadowContext(nullptr)
+{
+}
+
+W3DProjectedShadowManager::~W3DProjectedShadowManager()
+{
+    Release_Resources();
+    m_dynamicRenderTarget = nullptr;
+    m_renderTargetHasAlpha = false;
+
+    if (m_shadowContext) {
+        delete m_shadowContext;
+    }
+
+    if (m_shadowCamera) {
+        delete m_shadowCamera;
+    }
+
+    m_shadowCamera = nullptr;
+
+    if (m_W3DShadowTextureManager) {
+        delete m_W3DShadowTextureManager;
+    }
+
+    m_W3DShadowTextureManager = nullptr;
+
+    captainslog_dbgassert(!m_shadowList, "Destroy of non-empty projected shadow list");
+    captainslog_dbgassert(!m_decalList, "Destroy of non-empty projected decal list");
+}
+
+void W3DProjectedShadowManager::Reset()
+{
+    captainslog_dbgassert(!m_shadowList, "Reset of non-empty projected shadow list");
+    captainslog_dbgassert(!m_decalList, "Reset of non-empty projected decal list");
+    m_W3DShadowTextureManager->Free_All_Textures();
+}
+
+bool W3DProjectedShadowManager::Init()
+{
+    m_W3DShadowTextureManager = new W3DShadowTextureManager();
+    m_shadowCamera = new CameraClass();
+    m_shadowContext = new SpecialRenderInfoClass(*m_shadowCamera, SpecialRenderInfoClass::RENDER_SHADOW);
+    m_shadowContext->m_lightEnvironment = &m_shadowLightEnv;
+    return true;
+}
+
+bool W3DProjectedShadowManager::Re_Acquire_Resources()
+{
+#ifdef BUILD_WITH_D3D8
+    captainslog_dbgassert(!m_dynamicRenderTarget, "Acquire of existing shadow render target");
+
+    m_renderTargetHasAlpha = true;
+    m_dynamicRenderTarget = DX8Wrapper::Create_Render_Target(512, 512, WW3D_FORMAT_A8R8G8B8);
+
+    if (!m_dynamicRenderTarget) {
+        m_renderTargetHasAlpha = false;
+        m_dynamicRenderTarget = DX8Wrapper::Create_Render_Target(512, 512, WW3D_FORMAT_UNKNOWN);
+    }
+
+    IDirect3DDevice8 *dev = DX8Wrapper::Get_D3D_Device8();
+    captainslog_dbgassert(dev, "Trying to ReAquireResources on W3DProjectedShadowManager without device");
+    captainslog_dbgassert(!g_shadowDecalIndexBufferD3D, "ReAquireResources not released in W3DProjectedShadowManager");
+
+    if (dev->CreateIndexBuffer(2 * g_shadowDecalIndexSize,
+            D3DUSAGE_WRITEONLY | D3DUSAGE_DYNAMIC,
+            D3DFMT_INDEX16,
+            D3DPOOL_DEFAULT,
+            &g_shadowDecalIndexBufferD3D)
+        < S_OK) {
+        return false;
+    }
+
+    return g_shadowDecalVertexBufferD3D
+        || dev->CreateVertexBuffer(24 * g_shadowDecalVertexSize,
+               D3DUSAGE_WRITEONLY | D3DUSAGE_DYNAMIC,
+               0,
```

Comment:
D3DFMT_UNKNOWN?

Reply:
It takes an FVF so 0 is appropriate here (or one or more of the D3DFVF constants)